### DISCUSSION TITLE
Persistent state_set outside of pyscript domain?

### DISCUSSION THREAD

**Discussion by selmaohneh** [custom-components/pyscript#324] [Q&A] [answered]
Hi!

Just found out about pyscript yesterday - and I am already planning on moving all my automation from NodeRed to pyscript. :-D Thanks for all the work!

I create a lot of custom sensors with pyscript, just as a better alternative to template sensors. 
But since they need to persist state, all of them lie inside the pyscript.* domain. But I want a light entity to appear in the light.* domain. 

Check out this example:

```
state.persist(
    "pyscript.livingroom_candles", 
    default_value="off", 
    default_attributes={"device_class": "light", "icon": "mdi:candle"});

@service
def livingroom_candles_turnon():
    remote.livingroom_universalremote_remote.send_command(device="candleremote", command="on");
    updateSensor("on");

@service
def livingroom_candles_turnoff():
    remote.livingroom_universalremote_remote.send_command(device="candleremote", command="off");
    updateSensor("off");

@service
def livingroom_candles_toggle():
    if pyscript.livingroom_candles == "on":
        livingroom_candles_turnoff();
        return;

    if pyscript.livingroom_candles == "off":
        livingroom_candles_turnon();
        return;

@service
def livingroom_candles_disablecandlemode():
    remote.livingroom_universalremote_remote.send_command(device="candleremote", command="lightmode");

def updateSensor(value):
    state.set(
        "pyscript.livingroom_candles", 
        value,
        device_class="light", 
        icon="mdi:candle");
```

This works perfectly fine! But I'd like to use `light.livingroom_candles` instead of `pyscript.livingroom_candles`. Is this possible?

Thanks in advance!

**Comment by dlashua** [comment, marked as answer]
It is entirely possible (i.e. Home Assistant will allow it). However, the decision was made to limit it to the "pyscript" domain. Changes need to be made [here](https://github.com/custom-components/pyscript/blob/17df2e4c1ecc54e0209217199d8a0d5e69460adf/custom_components/pyscript/state.py#L222) to allow it to persist in any domain.

Edit: Keep in mind, persisting sensor and binary_sensor is safe and easy since there are no service calls to be made on them. However, persisting light, for instance, is problematic with a change like this. While the state updates will work, this will not handle service calls (i.e. light.turn_on). So the light entity it creates will not response to UI button presses, etc. The proper implementation for light, switch, climate, etc is to implement the Home Assistant setup methods for each of those platforms with hooks back into user land pyscript code to handle the service calls.

**Reply by selmaohneh** [reply]
Thanks for the clarification!

**Reply by selmaohneh** [reply]
Another thought: What about just the domain `sensor`?

This would not contain the complexity you mentioned, i.e. the need to work with a lot of domain services. 

But instead we would be able to use this sensor with all the new features that are dropping in the last months. I am thinking of the statistics sensor, utility meter or the energy dashboard. All these features need a `sensor` and don't allow any other domain...
